controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	// Create a provisioner to access Ironic API
+	prov, err := r.ProvisionerFactory.NewProvisioner(ctx, provisioner.BuildHostDataNoBMC(*bmh), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create provisioner: %w", err)
+	}
+
+	ready, err := prov.TryInit()
+	if err != nil || !ready {
+		var msg string
+		if err == nil {
+			msg = "not ready"
+		} else {
+			msg = err.Error()
+		}
+		reqLogger.Info("provisioner is not ready", "Error", msg, "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
+	}
+
+	newStatus, err := r.updateHostFirmware(info)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not update hostfirmwarecomponents: %w", err)
+	}
+
+	// Check ironic for the components information if possible
+	components, err := prov.GetFirmwareComponents()
+	info.log.Info("retrieving firmware components and saving to resource", "Node", bmh.Status.Provisioning.ID)
+
+	if err != nil {
+		reqLogger.Error(err, "provisioner returns error", "RequeueAfter", provisionerRetryDelay)
+		setUpdatesCondition(info.hfc.GetGeneration(), &newStatus, info, metal3api.HostFirmwareComponentsInvalid, metav1.ConditionTrue, reasonInvalidComponent, err.Error())
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, err
+	}
+
+	if err = r.updateHostFirmwareComponents(newStatus, components, info); err != nil {
+		return ctrl.Result{Requeue: false}, err
+	}
+
+	for _, e := range info.events {
+		r.publishEvent(info.ctx, req, e)
+	}
+
+	if meta.IsStatusConditionTrue(info.hfc.Status.Conditions, string(metal3api.HostFirmwareComponentsChangeDetected)) {
+		return ctrl.Result{Requeue: true, RequeueAfter: reconcilerRequeueDelayChangeDetected}, nil
+	}
+	return ctrl.Result{Requeue: true, RequeueAfter: reconcilerRequeueDelay}, nil
+}
+
+// Update the HostFirmwareComponents resource using the components from provisioner.
+func (r *HostFirmwareComponentsReconciler) updateHostFirmware(info *rhfcInfo) (newStatus metal3api.HostFirmwareComponentsStatus, err error) {
+	dirty := false
+
+	// change the Updates in Status
+	newStatus.Updates = info.hfc.Spec.Updates
+
+	// Check if the updates in the Spec are different than Status
+	updatesMismatch := !reflect.DeepEqual(info.hfc.Status.Updates, info.hfc.Spec.Updates)
+
+	reason := reasonValidComponent
+	generation := info.hfc.GetGeneration()
+
+	if updatesMismatch {
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsChangeDetected, metav1.ConditionTrue, reason, "") {
+			dirty = true
+		}
+
+		err := r.validateHostFirmwareComponents(info)
+		if err != nil {
+			info.publishEvent("ValidationFailed", fmt.Sprintf("Invalid Firmware Components: %s", err))
+			reason = reasonInvalidComponent
+			if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsValid, metav1.ConditionFalse, reason, fmt.Sprintf("Invalid Firmware Components: %s", err)) {
```

Comment:
This can be somewhat confusing. You have both conditions Valid and Invalid, both taking boolean values, used in different contexts. Maybe you can replace Invalid above with setting Valid to False?

Otherwise, you need to make sure you set Valid to False when setting Invalid to True and vice versa.

Reply:
you mean `reason` being `reasonValidComponent` or `reasonInvalidComponent` right? and switch to use just boolean `true or false`  ?

Reply:
I mean, you have both metal3api.HostFirmwareComponentsInvalid and metal3api.HostFirmwareComponentsValid. Since conditions are essentially booleans, these look redundant.

Reply:
Gotcha! updating now